lib/HLSL/HLOperationLower.cpp

```diff
@@ -459,17 +467,42 @@ Value *TrivialDxilOperation(Function *dxilFunc, OP::OpCode opcode,
     }
   }
 }
-// Generates a DXIL operation over an overloaded type (Ty), returning a
-// RetTy value; when Ty is a vector, it will replicate per-element operations
-// into RetTy to rebuild it.
-Value *TrivialDxilOperation(OP::OpCode opcode, ArrayRef<Value *> refArgs,
-                            Type *Ty, Type *RetTy, OP *hlslOP,
-                            IRBuilder<> &Builder) {
+
+// Creates a native vector call to for a "trivial" operation where only a single
+// call instruction is needed. The overload and return types are the same vector
+// type `Ty`.
+// Utility objects `HlslOp` and `Builder` are used to create a call to the given
+// `DxilFunc` with `RefArgs` arguments.
+Value *TrivialDxilVectorOperation(Function *Func, OP::OpCode Opcode,
+                                  ArrayRef<Value *> Args, Type *Ty, OP *OP,
+                                  IRBuilder<> &Builder) {
+  if (!Ty->isVoidTy())
+    return Builder.CreateCall(Func, Args, OP->GetOpCodeName(Opcode));
+  else
+    return Builder.CreateCall(Func, Args); // Cannot add name to void.
+}
+
+// Generates a DXIL operation with the overloaded type based on `Ty` and return
+// type `RetTy`. When Ty is a vector, it will either generate per-element calls
+// for each vector element and reconstruct the vector type from those results or
+// operate on and return native vectors depending on vector size and the value
+// of `SupportsVectors`, which is deteremined by version and opcode support.
+Value *TrivialDxilOperation(OP::OpCode Opcode, ArrayRef<Value *> Args, Type *Ty,
+                            Type *RetTy, OP *OP, IRBuilder<> &Builder,
+                            bool SupportsVectors = false) {
```

Comment:
Instead of the caller supplying `SupportsVectors`, couldn't this be looked up by the operation and the module?  Something like:

```cpp
if (Ty->isVectorTy() && Ty->getVectorNumElements() > 1 &&
    OP->GetModule()->GetHLModule()->GetShaderModel()->IsSM69Plus() &&
    OP::IsOverloadLegal(Opcode, Ty)) {
  // ...
}
```